test/functional/mempool_dump.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test mempool dump.
+
+By default, bitcoind will dump mempool on shutdown and
+then reload it on startup. Dump can also happen when requested via RPC.
+
+Test is as follows:
+
+  - start node0
+  - check no mempool file exist
+  - dump mempool and verify it then exists
+  - move parent directory to cause fail to dump, verify exception
+
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import *
+import os
+
+class MempoolDumpTest(BitcoinTestFramework):
+
+    def __init__(self):
+        super().__init__()
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+
+    def run_test(self):
+        self.log.debug("Verify mempool file doesn't exists")
+        mempooldat = self.options.tmpdir + "/node0/regtest/mempool.dat"
+        assert_raises(OSError, lambda: os.remove(mempooldat))
+        self.log.debug("Dump mempool to disk")
+        assert_equal(self.nodes[0].dumpmempool(), True)
+        assert_equal(os.path.isfile(mempooldat), True)
+        os.rename(self.options.tmpdir + "/node0/regtest", self.options.tmpdir + "/node0/regtest_bk")
```

Comment:
You might be able to achieve the same error condition by changing the file permissions, though.